File: src/main/java/io/github/sunshinewzy/shining/api/blueprint/IBlueprintNodeTree.java
package io.github.sunshinewzy.shining.api.blueprint;

import org.jetbrains.annotations.NotNull;

/**
 * The logical unit of blueprints
 */
public interface IBlueprintNodeTree {
	
	/**
	 * Gets the root of the node tree.
	 */
	@NotNull
	IBlueprintNode getRoot();

	/**
	 * Sets the root of the node tree.
	 *
	 * @return The previous root node
	 */
	@NotNull
	IBlueprintNode setRoot(@NotNull IBlueprintNode node);

}

File: src/main/kotlin/io/github/sunshinewzy/shining/api/blueprint/IBlueprintNodeTree.kt
package io.github.sunshinewzy.shining.api.blueprint

/**
 * The logical unit of blueprints
 */
interface IBlueprintNodeTree {

    /**
     * Gets the root of the node tree.
     */
    fun getRootOrNull(): IBlueprintNode?

    /**
     * Gets the root of the node tree.
     * Throws an exception when the root does not exist.
     */
    fun getRoot(): IBlueprintNode = getRootOrNull()!!

    /**
     * Sets the root of the node tree.
     * 
     * @return The previous predecessor node, or null if it does not exist
     */
    fun setRoot(node: IBlueprintNode): IBlueprintNode?
    
}